### DISCUSSION TITLE
Calling getPayout(<keyword>) before getOfferResult(): does it ever make sense?

### DISCUSSION THREAD

**Discussion by 0xshipthecode** [Agoric/agoric-sdk#5720] [Q&A] [answered]
it's actually possible to call `E(seat).getPayout(<keyword>)` before calling `E(seat).getOfferResult()`. The `getPayout()` call typically completes with a 0 value (this seems to depend on details of the offer handler). Also if the contract exits using e.g. an `assert` call, the exception is thrown in the callers context only after `getOfferResult()` is called. Thus one could completely miss the fact that the contract call has failed if `getPayout()` is called without `getOfferResult()` called first.

My question is this. Does it ever make sense to call `getPayout(...)` before calling `getOfferResult()`?

If not: would it make sense for `getPayout(...)` to throw an exception if `getOfferResult()` hasn't been called yet?

Either way it would be great to have guidance in docs on this

**Comment by 0xshipthecode** [comment]
I'm updating this comment to reflect new understanding from the code.

1. It seems important to call `getOfferRestul()` because that is how we should check whatever the contract has decided to do. Also it seems to be how exceptions get propagated into caller code.
3. However it's also seems important to verify that a seat `hasExited()` before calling `getPayout()`, since `getPayouts()` is simply going to have no payouts before that.

So it would seem that a standard sequence should be (assuming contract exists the seat in the handler on success):

```js
const result = await E(seat).getOfferResult()

if (isSuccessful(result)) {

  // make sure seat has actually exited
  assert(await E(seat).hasExited())

  // only now retrieve payouts
  const payouts = await E(seat).getPayouts()
}
```

Is this correct?

Obviously whether we should expect the seat to be exited depends on the contract. We may have to wait and check multiple times (or get notified), however those are more complex flows that aren't part of this issue.

**Comment by dckc** [comment, marked as answer]
@Chris-Hibbert tells me:

It's completely plausible to call `E.when(getPayout(), ...)`, without calling `getOfferResult()`. Zoe explicitly returns a promise there, and doesn't resolve the promise until the seat is exited.

**Comment by dckc** [comment]
Also, while some seats exit during their offer handler, others may stay open for, for example, the duration of an auction. So `assert(await E(seat).hasExited())` isn't guaranteed after `E(seat).getOfferResult()`.
